### ISSUE TITLE
MAP metric breaks method `state_dict` from `state.py`

### ISSUE THREAD

**Issue by priba** [mosaicml/composer#2471]
Composer version `0.16.0` breaks when eval metrics is `composer.MAP`.

Note that MAP provides a dictionary of Tensors and we cannot get its device in:

https://github.com/mosaicml/composer/blob/934732f14d5294be55c1e20a2db35d697c63695c/composer/core/state.py#L925

**Comment by mvpatel2000** [comment]
Thanks for flagging this! We will do our best to patch and release a hotfix asap.

**Comment by priba** [comment]
Hi @eracah! Thanks for taking this.

This issue is strongly related to #2413

I've been thinking of the best solution and it seems that replacing the `.compute()` assignments like:

```python
for metric_name, metric in metrics.items():
    computed_metrics[metric_name] = metric.compute()
```

See an example here, although it happen in several places:

https://github.com/mosaicml/composer/blob/a83beab8365ec5c9a5ea6b7201a842b599b82c37/composer/trainer/trainer.py#L1910 

With something like:

```python
 def flatten_metrics(metric_name: str, metric_value: Any) -> dict[str, Any]: 

    if isinstence(metric_value, Tensor): # Maybe also number?
        return {metric_name: metric_value}

    flattened_dict = {}
    if isinstance(metric_value, list): # list/sequence/iterable etc.
          for i, v in enumerate(metric_value):
              flattened_dict.update(flatten_metrics(f"{metric_name}_{i}", v))
    elif isinstance(metric_value, dict):
          for l, v in metric_value.items():
              flattened_dict.update(flatten_metrics(f"{metric_name}/{l}], v))
    return {metric_name: metric_value}

....

for metric_name, metric in metrics.items():
    metric_value = metric.compute()
    computed_metrics.update(flatten_metrics(metric_name, metric_value)) # Maybe raise an error if two keys are repeated?
```

Note that this is pseudocode and has not been tested.

**Comment by mvpatel2000** [comment]
This should be fixed in 0.16.1!

Long answer short, torchmetrics has tensors which get sharded and do not play well with sharded checkpoints. We cast them to numpy to avoid this, but this seems to have introduced a host of problems. In 0.16.1, for the normal case we reverted back to the prior path and for sharded checkpointing emit a warning + do not save metrics. 

In general, metrics seem very hard to save and have often caused compatibility issues. In the future we plan on not saving metrics at all into checkpoints unless explicitly toggled on with a flag.

I'm going to close this issue for now, but please feel free to reopen if problems arise. I've added a unit test for MAP specifically to avoid this in the future

**Comment by priba** [comment]
It seems it works fine. Thanks @mvpatel2000
